pyomo/contrib/doe/fim_doe.py

```diff
@@ -0,0 +1,2582 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#
+#  Pyomo.DOE was produced under the DOE Carbon Capture Simulation 
+#  Initiative (CCSI), and is copyright (c) 2022 by the software owners: 
+#  TRIAD, LLNS, BERKELEY LAB, PNNL, UT-Battelle, LLC, NOTRE
+#  DAME, PITT, UT Austin, TOLEDO, WVU, et al. All rights reserved.
+# 
+#  NOTICE. This Software was developed under funding from the 
+#  U.S. Department of Energy and the U.S. Government consequently retains 
+#  certain rights. As such, the U.S. Government has been granted for itself
+#  and others acting on its behalf a paid-up, nonexclusive, irrevocable, 
+#  worldwide license in the Software to reproduce, distribute copies to the 
+#  public, prepare derivative works, and perform publicly and display
+#  publicly, and to permit other to do so.
+#  ___________________________________________________________________________
+
+from pyomo.common.dependencies import (
+    numpy as np, numpy_available,
+    pandas as pd, pandas_available,
+    scipy, scipy_available,
+    matplotlib as plt, matplotlib_available,
+)
+
+import pyomo.environ as pyo
+from pyomo.opt import SolverFactory, SolverStatus, TerminationCondition
+import time
+import pickle
+from itertools import permutations, product
+import logging
+from pyomo.contrib.sensitivity_toolbox.sens import sipopt, sensitivity_calculation, get_dsdp
+
+class Measurements:
+    def __init__(self, measurement_index_time, variance=None, ind_string='_index_'):
+        """This class stores measurements' information
+
```

Comment:
Replace the "This class... information" with the following description.

This class stores information on which algebraic and differential variables in the Pyomo model are considered measurements. This includes the functionality to specify indices for these measurement variables. For example, with a partial differential algebraic equation model, these measurement index sets can specify which spatial and temporal coordinates each measurement is available. Moreover, this class supports defining the covariance matrix for all measurements.